src/nccl_ofi_net.c

```diff
@@ -1050,49 +1064,130 @@ static inline ncclResult_t nccl_ofi_progress(nccl_ofi_t *nccl_ofi_comp)
 }
 
 /*
- * @brief	Checks if the underlying hardware type matches the input platform type.
+ * @brief	Provides EC2 platform type as reported by the
+ * 		first line of
+ *		/sys/devices/virtual/dmi/id/product_name.
+ *		Users of this API *should* free the buffer when a
+ *		Non-NULL string is returned.
  *
- * @return	-1, on error
- *		1, if it matches
- *		0, if it does not match
+ * @return	NULL, on allocation and file system error
+ * 		EC2 platform type, on success
  */
-static int platform_type_matches(char *platform_type)
+static char* get_platform_type()
 {
 	char file[] = "/sys/devices/virtual/dmi/id/product_name";
 	FILE *fd = NULL;
-	int len = strlen(platform_type);
-	int idx = 0, ret = -1;
 	char ch;
+	size_t len = 0;
+	size_t platform_type_len = 64;
+	char *platform_type = NULL;
 
 	fd = fopen(file, "r");
 	if (fd == NULL) {
 		NCCL_OFI_WARN("Error opening file: %s", file);
-		return -1;
+		goto error;
 	}
 
-	/* Read first line of the file */
+	platform_type = (char *)malloc(sizeof(char)*platform_type_len);
+	if (platform_type == NULL) {
+		NCCL_OFI_WARN("Unable to allocate platform type");
+		goto error;
+	}
+
+	/* Read first line of the file, reallocing the buffer as necessary */
 	while ((feof(fd) == 0) && (ferror(fd) == 0) && ((ch = fgetc(fd)) != '\n')) {
+		platform_type[len++] = ch;
+		if (len > platform_type_len) {
```

Comment:
I think this is off by 1?  

Imagine `len == 63` at the start of the loop iteration.  We do `platform_type[63] = ch`, then the post-increment increases `len` to 64.  `(64 > 64) == FALSE` so we loop around and assign `platform_type[64] = ch`, boom, buffer overflow.

Reply:
fixed